[ISS-5465] add a flag to let CA delegate node deletion to the external node provisioner

wmgroot opened the pull request:
#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

damienleger committed:
Add container securityContext

- container is now immutable (file integrity guaranteed): readOnlyRootFilesystem
- remove rights the application doesn't need: allowPrivilegeEscalation+capabilities

damienleger committed:
Update cluster-autoscaler/cloudprovider/aws/examples/cluster-autoscaler-autodiscover.yaml

Co-authored-by: Joe Bowbeer <joe.bowbeer@gmail.com>

damienleger committed:
Update cluster-autoscaler/cloudprovider/aws/examples/cluster-autoscaler-multi-asg.yaml

Co-authored-by: Joe Bowbeer <joe.bowbeer@gmail.com>

damienleger committed:
Update cluster-autoscaler/cloudprovider/aws/examples/cluster-autoscaler-one-asg.yaml

Co-authored-by: Joe Bowbeer <joe.bowbeer@gmail.com>

damienleger committed:
Update cluster-autoscaler/cloudprovider/aws/examples/cluster-autoscaler-run-on-control-plane.yaml

Co-authored-by: Joe Bowbeer <joe.bowbeer@gmail.com>

damienleger committed:
add seccompProfile

Commit:
Update Ionos Cloud README and deploy example

domenicbozzuto committed:
Fix scaledown:nodedeletion metric calculation

The scaledown:nodedeletion metric duration was incorrectly being computed relative to the start of the RunOnce routine, instead of from the actual start of the deletion. Behavior in the start of the routine (like a long cloudproviderrefresh) would incorrectly skew the nodedeletion duration

Signed-off-by: Domenic Bozzuto <dom.bozzuto@datadoghq.com>

Commit:
cluster-autoscaler: Add option to disable scale down of unready nodes

Add flag '--scale-down-unready-enabled' to enable or disable scale-down
of unready nodes. Default value set to true for backwards compatibility
(i.e., allow scale-down of unready nodes).

Signed-off-by: Grigoris Thanasoulas <gregth@arrikto.com>

Shubham82 committed:
Added RBAC Permission to cherryservers.

wmgroot committed:
[ISS-5465] add a flag to let CA delegate node deletion to the external node provisioner

Signed-off-by: wmgroot <wmgroot@gmail.com>

mmcclean-aws committed:
Added support for in2 instance types

Added support for the AWS Inferentia 2 instance types based on the NeuronCore v2 chip architecture

mmcclean-aws committed:
added inf2 instance types to ec2 api.go and api-2.json files

kushagra98 committed:
upgrade vpa and vpa-e2e vendor to 1.26

kushagra98 committed:
fix tests due to vendor upgrade

kushagra98 committed:
add script to upgrade vpa vendor

cnmcavoy commented on cluster-autoscaler/core/static_autoscaler.go:
```diff
@@ -800,6 +807,13 @@ func (a *StaticAutoscaler) deleteCreatedNodesWithErrors() (bool, error) {
+			} else {
+				err = nodeGroup.DeleteNodes(nodesToBeDeleted)
+			}
 			err = nodeGroup.DeleteNodes(nodesToBeDeleted)
```
Bug: line L817 needs to be removed, it duplicates L815.

wmgroot replied:
Unit tests actually caught this. Thanks!

cnmcavoy commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go:
```diff
@@ -88,6 +88,80 @@ func (ng *nodegroup) IncreaseSize(delta int) error {
+		return err
+	}
+
+	// if we are at minSize already we wail early.
```
nit: typo, 'fail'

cnmcavoy commented on cluster-autoscaler/cloudprovider/cloud_provider.go:
```diff
@@ -176,6 +176,11 @@ type NodeGroup interface {
 	// node group size is updated. Implementation required.
 	IncreaseSize(delta int) error
 
+	// MarkNodesForDeletion tells the cloud provider to delete the nodes from this node group.
```
This comment doesn't fully capture how `MarkNodesForDeletion` is different from `DeleteNodes`. I think we should  update the comment on `DeleteNodes` as well, to something like this:
```
	// DeleteNodes deletes nodes from this node group and shrinks the node
	// group target size by the number of deleted nodes. Error is returned either on
	// failure or if the given node doesn't belong to this node group. This function
	// should wait until node group size is updated. Implementation required.
```

cnmcavoy commented on cluster-autoscaler/config/autoscaling_options.go:
```diff
@@ -247,4 +247,6 @@ type AutoscalingOptions struct {
 	ParallelDrain bool
 	// NodeGroupSetRatio is a collection of ratios used by CA used to make scaling decisions.
 	NodeGroupSetRatios NodeGroupDifferenceRatios
+	// ExternalNodeDeletion configures CA to let the external cloud providers handle deletion of nodes.
```
This comment does not explain when or why a user might want to enable this setting. We should expand this to clarify why you might want to enable this and what providers support it right now (clusterapi).

cnmcavoy commented on cluster-autoscaler/main.go:
```diff
@@ -221,6 +221,7 @@ var (
 	maxFreeDifferenceRatio             = flag.Float64("max-free-difference-ratio", config.DefaultMaxFreeDifferenceRatio, "Maximum difference in free resources between two similar node groups to be considered for balancing. Value is a ratio of the smaller node group's free resource.")
 	maxAllocatableDifferenceRatio      = flag.Float64("max-allocatable-difference-ratio", config.DefaultMaxAllocatableDifferenceRatio, "Maximum difference in allocatable resources between two similar node groups to be considered for balancing. Value is a ratio of the smaller node group's allocatable resource.")
 	forceDaemonSets                    = flag.Bool("force-ds", false, "Blocks scale-up of node groups too small for all suitable Daemon Sets pods.")
+	externalNodeDeletion               = flag.Bool("external-node-deletion", false, "If true, CA will delegate node deletion to each cloud provider.")
```
Same comment as in the `autoscaling_options.go`.

cnmcavoy commented:
@wmgroot there is one other `DeleteNodes` usage, in `deleteNodesFromCloudProvider` in `cluster-autoscaler/core/scaledown/actuation` in `delete_in_batch.go`:
```
func deleteNodesFromCloudProvider(ctx *context.AutoscalingContext, nodes []*apiv1.Node) (cloudprovider.NodeGroup, error) {
	nodeGroup, err := ctx.CloudProvider.NodeGroupForNode(nodes[0])
	if err != nil {
		return nodeGroup, errors.NewAutoscalerError(errors.CloudProviderError, "failed to find node group for %s: %v", nodes[0].Name, err)
	}
	if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
		return nodeGroup, errors.NewAutoscalerError(errors.InternalError, "picked node that doesn't belong to a node group: %s", nodes[0].Name)
	}
	if err = nodeGroup.DeleteNodes(nodes); err != nil {
		return nodeGroup, errors.NewAutoscalerError(errors.CloudProviderError, "failed to delete %s: %v", nodes[0].Name, err)
	}
	return nodeGroup, nil
}
```
I think we want to handle this different than the others, since the scale down in the target size is expected. Thoughts?

wmgroot commented:
> @wmgroot there is one other `DeleteNodes` usage, in `deleteNodesFromCloudProvider` in `cluster-autoscaler/core/scaledown/actuation` in `delete_in_batch.go`:
> 
> ```
> func deleteNodesFromCloudProvider(ctx *context.AutoscalingContext, nodes []*apiv1.Node) (cloudprovider.NodeGroup, error) {
> 	nodeGroup, err := ctx.CloudProvider.NodeGroupForNode(nodes[0])
> 	if err != nil {
> 		return nodeGroup, errors.NewAutoscalerError(errors.CloudProviderError, "failed to find node group for %s: %v", nodes[0].Name, err)
> 	}
> 	if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
> 		return nodeGroup, errors.NewAutoscalerError(errors.InternalError, "picked node that doesn't belong to a node group: %s", nodes[0].Name)
> 	}
> 	if err = nodeGroup.DeleteNodes(nodes); err != nil {
> 		return nodeGroup, errors.NewAutoscalerError(errors.CloudProviderError, "failed to delete %s: %v", nodes[0].Name, err)
> 	}
> 	return nodeGroup, nil
> }
> ```
> 
> I think we want to handle this different than the others, since the scale down in the target size is expected. Thoughts?

My thinking was to keep the downscaling implementation unchanged, since modifying the replica count is intended scale down behavior. I might be missing your point though.

cnmcavoy commented:
> My thinking was to keep the downscaling implementation unchanged, since modifying the replica count is intended scale down behavior. I might be missing your point though.

I guess I am trying to consider this PR from an external CAPI user's perspective. Would they be confused that the with this new CA "mode" that it mixes both behaviors? It makes the intent of this change seem unclear.

pbetkier committed:
Fix multiple recommenders example: deployment name

Seems like the example in `/vertical-pod-autoscaler/deploy/recommender-deployment-low.yaml` copied `Deployment` name from `recommender-deployment-high.yaml`.

jbartosik committed:
Create enhancements directory for Addon Resizer

k8s-ci-robot committed:
Merge pull request #5512 from Shubham82/add_RBAC_permissions_cherryservers

Added RBAC Permission to cherryservers.

k8s-ci-robot committed:
Merge pull request #5472 from DataDog/scaledown-nodedeletion-metric-fix

Fix scaledown:nodedeletion metric calculation

yaroslava-serdiuk committed:
Fix RemovalSimulation for parallel scale down

mboersma committed:
cluster-autoscaler: add missing import to kubermark_other.go

k8s-ci-robot committed:
Merge pull request #5539 from pbetkier/patch-1

Fix multiple recommenders example: deployment name

yaroslava-serdiuk committed:
Add limit for removable nodes count

k8s-ci-robot committed:
Merge pull request #5552 from yaroslava-serdiuk/scalability

Fix RemovalSimulation for parallel scale down

x13n committed:
Limit the rate of adding new uneeded nodes

This is an optimization to avoid spending too much time in scale down
simulation. The main idea is that it doesn't make sense to add new
unneeded nodes at a rate that is higher than the rate at which we can
delete them afterwards.

k8s-ci-robot committed:
Merge pull request #5556 from x13n/fast-sd

Limit the rate of adding new uneeded nodes

peaaceChoi committed:
Delete unused return param

peaaceChoi committed:
Fix continue condition

yaroslava-serdiuk committed:
Add Snapshot() method to the NodeDeletionTracker

k8s-ci-robot committed:
Merge pull request #5562 from yaroslava-serdiuk/deletions-in-progress

Use snapshot of NodeDeletionTracker for ActuationStatus

k8s-ci-robot committed:
Merge pull request #5559 from ionos-cloud/update-ionos-cloud-docs

Update Ionos Cloud README and deploy example

k8s-ci-robot committed:
Merge pull request #5561 from peaaceChoi/remove-deepcopynode-err

Delete unused err  returns

k8s-ci-robot committed:
Merge pull request #4478 from damienleger/container_security_context

[CA] [AWS examples] Add container securityContext

mmcclean-aws committed:
Merge remote-tracking branch 'upstream/master'
Merged

mmcclean-aws committed:
rolled back VPA files

k8s-ci-robot committed:
Merge pull request #5549 from mboersma/fix-gpu-import

cluster-autoscaler: add missing import to kubermark_other.go

k8s-ci-robot committed:
Merge pull request #5545 from jbartosik/addon-resizer-kep

Create enhancements directory for Addon Resizer

qiutongs committed:
Cluster Autoscaler: GCE: add pricing info for C3

m99coder committed:
add `hostNetwork` parameter

k8s-ci-robot committed:
Merge pull request #5572 from qiutongs/master

Cluster Autoscaler: GCE: add pricing info for C3

Commit:
docs(README): Update information about EKS PrivateLink

Since EKS API now supports PrivateLink, update the AWS README with the
current state.

Signed-off-by: Grigoris Thanasoulas <gregth@arrikto.com>

voelzmo committed:
Remove unused OOM Bump constants

voelzmo committed:
Switch to built-in Duration.truncate method

k8s-ci-robot committed:
Merge pull request #5463 from m99coder/patch-1

Re-sort on root level, add `hostNetwork` parameter

yaroslava-serdiuk committed:
Add empty nodes sorting for scale down candidates

k8s-ci-robot committed:
Merge pull request #5537 from arrikto/feature-disable-unready-scaledown

cluster-autoscaler: Add option to disable scale down of unready nodes

mmcclean-aws committed:
Removed whitespace

Co-authored-by: Guy Templeton <guyjtempleton@googlemail.com>

k8s-ci-robot committed:
Merge pull request #5529 from mmcclean-aws/master

Added support for AWS inf2 instance types

k8s-ci-robot committed:
Merge pull request #5575 from voelzmo/enh/cleanup-recommender

Cleanup recommender model/container.go

k8s-ci-robot committed:
Merge pull request #5574 from arrikto/feature-eks-readme

docs(README): Update information about EKS PrivateLink

k8s-ci-robot committed:
Merge pull request #5578 from yaroslava-serdiuk/empty-sorting

Add empty nodes sorting for scale down candidates

gjtempleton committed:
CA - Document Debugging Snapshotter flag

k8s-ci-robot committed:
Merge pull request #5536 from kushagra98/update-vendor-K8s-1.26

Update vendor k8s 1.26

k8s-ci-robot committed:
Merge pull request #5581 from gjtempleton/CA-Document-Snapshotter-Flag

CA - Document Debugging Snapshotter flag
